components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/OAuthAppTenantResolverValve.java

```diff
@@ -0,0 +1,143 @@
+import org.wso2.carbon.identity.oauth.dao.OAuthAppDO;
+import org.wso2.carbon.identity.oauth2.IdentityOAuth2Exception;
+import org.wso2.carbon.identity.oauth2.client.authentication.OAuthClientAuthnException;
+import org.wso2.carbon.identity.oauth2.util.OAuth2Util;
+
+import java.io.IOException;
+
+import javax.servlet.ServletException;
+
+import static org.wso2.carbon.identity.oauth.common.OAuthConstants.OAuth10AParams.OAUTH_CONSUMER_KEY;
+import static org.wso2.carbon.identity.oauth.common.OAuthConstants.OAuth20Params.CLIENT_ID;
+import static org.wso2.carbon.identity.oauth.common.OAuthConstants.TENANT_NAME_FROM_CONTEXT;
+
+/**
+ * This valve is used to resolve the tenant domain of the OAuth application.
+ *
+ * When tenant qualified urls are not enabled, we need to set the tenant domain of the oauth app to the thread
+ * local. This is because with the client id tenant uniqueness improvement, DAO layer requires the tenant domain
+ * and client id to retrieve an app and the tenant is not available in the request path. Note that when tenant
+ * qualified urls are disabled, client id is unique across the server.
+ */
+public class OAuthAppTenantResolverValve extends ValveBase {
+
+    private static final Log log = LogFactory.getLog(OAuthAppTenantResolverValve.class);
+
+    @Override
+    public void invoke(Request request, Response response) throws IOException, ServletException {
+
+        try {
+            if (!IdentityTenantUtil.isTenantQualifiedUrlsEnabled() && isOAuthRequest(request)) {
+                String appTenant = "";
+                String clientId = request.getParameter(CLIENT_ID);
+
+                if (StringUtils.isEmpty(clientId) && isOAuth10ARequest(request)) {
+                    clientId = request.getParameter(OAUTH_CONSUMER_KEY);
+                }
+
+                // If empty, try to get the client id from the authorization header.
+                if (StringUtils.isEmpty(clientId)) {
+                    try {
+                        String[] credentials = OAuth2Util.extractCredentialsFromAuthzHeader(request);
+                        if (credentials.length == 2 && StringUtils.isNotEmpty(credentials[0])) {
+                            clientId = credentials[0];
+                        }
+                    } catch (OAuthClientAuthnException e) {
+                        if (log.isDebugEnabled()) {
+                            log.debug("Error while extracting credentials from authorization header.", e);
+                        }
+                    }
+                }
+
+                if (StringUtils.isNotEmpty(clientId)) {
+                    try {
+                        OAuthAppDO oAuthAppDO = OAuth2Util.getAppInformationByClientIdOnly(clientId);
+                        appTenant = OAuth2Util.getTenantDomainOfOauthApp(oAuthAppDO);
+                    } catch (IdentityOAuth2Exception e) {
+                        log.error("Error while getting oauth app for client Id: " + clientId, e);
```

Comment:
In the case of client ID getting resolved from authz header, the flow will be similar to a user's credential based authz as well, right?

If that's the case, for user authz requests also we will try to find a client with the username, and log an error, which will trigger false alarms.

Reply:
If a client application is not found, it will be a `InvalidOAuthClientException` which will be caught in the other catch block and log as a debug.